File: Rust/2.rs
fn main(args: ~[str]) {
  let mut sum = 0;

  let mut i = 0;
  let mut j = 1;

  while j < 4000000 {
    if j % 2 == 0 {
      sum += j;
    }
    let temp = j;
    j = j + i;
    i = temp;
  }

  io::println(int::str(sum));
}

File: go/2.go
package main

func main() {
	var sum int64 = 0
	var max int64 = 4000000
	fib := fib_generator()

	for i := fib(); i < max; i = fib() {
		if i%2 == 0 {
			sum += i
		}
	}

	println(sum)
}

func fib_generator() func() int64 {
	var prev int64 = 0
	var curr int64 = 1
	return func() int64 {
		temp := curr
		curr = curr + prev
		prev = temp
		return curr
	}
}
